Improving preemption performance

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
Scheduler takes a lot of time to find out the run job that it needs to preempt in order to run a high priority job. This process is slow and this pull request covers that.

#### Affected Platform(s)
All

#### Cause / Analysis / Design
The function select_index_to_preempt() does a lot of checking to see why the higher priority job did not run and then use to check for each running job (in case of hard limits). This comparison IMO took a long time and slowed preemption down.

#### Solution Description
Scheduler now culls the list of running jobs before it calls select_index_to_preempt and string comparisons have now been removed from select_index_to_preempt to speed up preemption.

Another place where scheduler used to spend a lot of time was in dup_node_info() function. Where it tries to find the job running on the node. With this change, now scheduler will store job/reservation's index from its array of all resource reservations and use that to find out the job, instead of going through all the elemets of the array and matching rank. This is implementation is similar to how nodes are found in unordered_nodes array.

#### Testing logs/output
[test_out.txt](https://github.com/PBSPro/pbspro/files/2413230/test_out.txt)



#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -5231,3 +5036,223 @@ extend_soft_walltime(resource_resv *resresv, time_t server_time)
+				return 1;
+			break;
+		case SERVER_JOB_LIMIT_REACHED:
+		case SERVER_RESOURCE_LIMIT_REACHED:
```
I don't think there is anything to compare here. This is the overall generic limit.

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -5231,3 +5036,223 @@ extend_soft_walltime(resource_resv *resresv, time_t server_time)
+			for (; job->select->chunks[index] != NULL; index++)
+			{
+				req_scan=job->select->chunks[index]->req;
+				while (req_scan)
```
This was the way it was before, I'll change it.

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -5231,3 +5036,223 @@ extend_soft_walltime(resource_resv *resresv, time_t server_time)
+ * @brief   This function looks at a list of running jobs and create a subset of preemptable candidates
+ *	    according to the high priority job and the reason why it couldn't run
+ * @param[in] arr - List of running jobs (preemptable candidates)
+ * @param[in] arr_length - number of preemptable candidates
```
Since the caller already has access to how bit the array would be, I just thought calculating it again would be an overhead.

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -5231,3 +5036,223 @@ extend_soft_walltime(resource_resv *resresv, time_t server_time)
+ *	    according to the high priority job and the reason why it couldn't run
+ * @param[in] arr - List of running jobs (preemptable candidates)
+ * @param[in] arr_length - number of preemptable candidates
+ * @param[in] job - high priority job that couldn't
```
oops :)

arungrover commented on src/scheduler/limits.c:
```diff
@@ -1365,9 +1366,11 @@ check_server_max_user_res(server_info *si, queue_info *qi, resource_resv *rr,
 			return (0);
 		case 1:	/* generic user limit exceeded */
 			schderr_args_server_res(NULL, resname, err);
```
You are right, I should remove it now that I'm passing resource definition.

arungrover commented on src/scheduler/limits.c:
```diff
@@ -2712,7 +2728,7 @@ check_max_group_res_soft(resource_resv *rr, counts *cts_list, void *limitctx)
  */
 static int
-check_max_user_res(resource_resv *rr, counts *cts_list, char **resname,
+check_max_user_res(resource_resv *rr, counts *cts_list, char **resname, resdef **rdef,
```
You are right, I'll modify that code.

arungrover commented on src/scheduler/node_info.c:
```diff
@@ -1479,8 +1479,8 @@ collect_jobs_on_nodes(node_info **ninfo_arr, resource_resv **resresv_arr, int si
-					if (find_resource_resv_by_rank(ninfo_arr[i]->job_arr,
-						job->rank) == NULL) {
+					if (find_resource_resv_by_indrank(ninfo_arr[i]->job_arr,
+						job->rank, job->resresv_ind) == NULL) {
```
Will do so.

arungrover commented on src/scheduler/server_info.c:
```diff
@@ -1861,6 +1861,64 @@ update_server_on_end(status *policy, server_info *sinfo, queue_info *qinfo,
+	}
+	nsinfo->jobs = job_arr;
+	nsinfo->all_resresv = all_arr;
+	nsinfo->sc.total = osinfo->sc.total;
```
you are right! But this was happening before as well in create_server_arrays() when dup_server_info called it. I guess it was not needed. I'll remove it.

arungrover commented on src/scheduler/simulate.c:
```diff
@@ -1772,7 +1772,7 @@ simulate_resmin(schd_resource *reslist, time_t end, event_list *calendar,
 		te = find_next_timed_event(te, IGNORE_DISABLED_EVENTS, event_mask)) {
 		resresv = (resource_resv *) te->event_ptr;
-		if (incl_arr == NULL || find_resource_resv_by_rank(incl_arr, resresv->rank) !=NULL) {
+		if (incl_arr == NULL || find_resource_resv_by_indrank(incl_arr, resresv->rank, resresv->resresv_ind) !=NULL) {
```
oh ok, I'll not use index here then

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,448 @@
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+import time
```
Got it! I'll remove it

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,448 @@
+        a = {'resources_available.ncpus': 4800,
+             'resources_available.mem': '2800mb'}
+        self.server.create_vnodes('vn', a, 1, self.mom, usenatvnode=True)
+        p = '"express_queue, normal_jobs, queue_softlimits, server_softlimits"'
```
Done!

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,448 @@
+        rv = self.server.manager(MGR_CMD_SET, QUEUE, a, 'workq', expect=True)
+        self.assertTrue(rv)
+
+        a = {'max_run_res.mem': "[u:pbsuser=1500mb]"}
```
Done

arungrover commented on src/scheduler/server_info.h:
```diff
@@ -143,6 +143,11 @@ update_server_on_run(status *policy, server_info *sinfo, queue_info *qinfo,
+/*
+ *	copy_server_arrays - copy server's jobs and all_resresv arrays
+ */
+int copy_server_arrays(server_info * nsinfo, server_info * osinfo);
```
will do

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,448 @@
+        self.assertTrue(rv)
+
+        a = {
+            'queue_type': 'e',
```
Done

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,448 @@
+        # Add qlist to the resources scheduler checks for
+        resources = self.scheduler.sched_config['resources']
+        resources = resources[:-1] + ', qlist\"'
+        self.scheduler.set_sched_config({'resources': resources})
```
Cool! I'll use that

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,448 @@
+
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+        jid = self.server.submit(j)
+        jid2 = self.server.submit(j2)
```
RIght, that can be done

arungrover commented:
Thanks for reviewing @bhroam. I've addressed your comments, please have a look.

arungrover commented:
[test_out.txt](https://github.com/PBSPro/pbspro/files/2440142/test_out.txt)

bhroam commented on src/scheduler/limits.c:
```diff
@@ -1423,11 +1420,10 @@ check_server_max_group_res(server_info *si, queue_info *qi, resource_resv *rr,
 			return (SERVER_GROUP_RES_LIMIT_REACHED);
 		case 2:	/* individual group limit exceeded */
-			schderr_args_server_res(rr->group, resname, err);
+			schderr_args_server_res(rr->group, rdef->name, err);
```
Why copy the rdef name into a string buffer instead of just using it directly in translate_fail_code().  The function already uses rdef directly.

arungrover replied:
Well I thought that resname was kind of redundant now that we have resource definition. I'm just passing it for completeness of the structure here, especially because group will still be part of error structure

arungrover replied:
If you feel strongly about it then let me know and I'll change it

bhroam commented on src/scheduler/limits.c:
```diff
@@ -1624,11 +1619,11 @@ check_queue_max_user_res(server_info *si, queue_info *qi, resource_resv *rr,
 			return (QUEUE_USER_RES_LIMIT_REACHED);
 		case 2:	/* individual user limit exceeded */
-			schderr_args_q_res(qi->name, rr->user, resname, err);
+			schderr_args_q_res(qi->name, rr->user, rdef->name, err);
```
Same here, use rdef directly (and other places).

bhroam commented on src/scheduler/server_info.c:
```diff
@@ -1899,23 +1899,22 @@ copy_server_arrays(server_info *nsinfo, server_info *osinfo)
-	qinfo = nsinfo->queues;
-	while (*qinfo != NULL) {
+
+	for (qinfo = nsinfo->queues; *qinfo != NULL; qinfo++) {
```
I know this code is preexisting, but if you want to be more consistent with the rest of the scheduler use an index instead of ++ing the array.  It's fine if you want to leave it alone.

arungrover replied:
will do

bhroam commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -61,78 +59,67 @@ def test_t1(self):
         for _ in range(7):
-            j = Job(TEST_USER1)
             a = {'Resource_List.select': '1:ncpus=3:mem=90mb',
                  'Resource_List.walltime': 9999}
```
a hasn't changed from above, no need to reset it.

bhroam commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -166,78 +153,67 @@ def test_t2(self):
         for _ in range(8):
-            j = Job(TEST_USER)
             a = {'Resource_List.select': '1:ncpus=3:mem=90mb',
                  'Resource_List.walltime': 9999}
```
a doesn't change in the loop, move up top.

arungrover replied:
Will do

bhroam commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -166,78 +153,67 @@ def test_t2(self):
         for _ in range(800):
-            j = Job(TEST_USER1)
             a = {'Resource_List.select': '1:ncpus=3',
                  'Resource_List.walltime': 9999}
```
a doesn't change in the loop, move out up top.

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -1869,7 +1861,18 @@ translate_fail_code(schd_error *err, char *comment_msg, char *log_msg)
 			}
 			break;
-			/* codes using a resource and arg1  */
+
```
Will do

arungrover commented:
@bhroam I've addressed your review comments

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -1872,45 +1879,39 @@ translate_fail_code(schd_error *err, char *comment_msg, char *log_msg)
 			break;
 
-			/* codes using three arguments (in a weird order) */
+		/* codes using three arguments (in a weird order) */
```
This isn't 3 args any more.  It's arg1, arg3, and the resource

arungrover commented:
@bhroam I've modified the comment. I also fixed a bug in copy_server_arrays function. It was an oversight from my side. Please have a look

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -5231,3 +5029,226 @@ extend_soft_walltime(resource_resv *resresv, time_t server_time)
+ * @retval - 0 if job is not valid for preemption
+ * @retval - 1 if the job is valid for preemption
+ */
+static int cull_running_jobs(resource_resv *job, void *arg)
```
How about cull_preemptible_jobs?

bhroam replied:
I like it :)

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -3188,8 +3198,7 @@ find_jobs_to_preempt(status *policy, resource_resv *hjob, server_info *sinfo, in
 	npolicy = nsinfo->policy;
-
-	njob = find_resource_resv_by_rank(nsinfo->jobs, hjob->rank);
+	njob = find_resource_resv_by_indrank(nsinfo->jobs, hjob->rank, hjob->resresv_ind);
```
This was existing code but for ease of reading I'll change it

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -3765,26 +3584,8 @@ select_index_to_preempt(status *policy, resource_resv *hjob,
-					 * RASSN resources are accumulated across the select.
-					 * This means all jobs will have them, and it invalidates the earlier check.
-					 */
-						if (resdef_exists_in_array(policy->resdef_to_check, req->def) &&
```
We perform a similar check in the same function below. Where we see if a high priority job is unable to find any good node then we try looking for server/queue resources.

arungrover replied:
I moved it to cull_preemptible_job in commit - d93dc3f.
It is now at line 5156 in job_info.c

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -3862,19 +3663,16 @@ select_index_to_preempt(status *policy, resource_resv *hjob,
 		}
 
-		if (!certainlygood && node_good == 0 && svr_res_good == 0)
+		if (node_good == 0 && svr_res_good == 0)
```
If we do not find a node to be good, we will reset this flag to 0 and search for server/queue level resources. So this flag will not be 0 always

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -3347,40 +3345,31 @@ find_jobs_to_preempt(status *policy, resource_resv *hjob, server_info *sinfo, in
-					/* don't forget current job found has been removed from sinfo->running_jobs! */
-					/* So we need to start again "where we last were" */
-						skipto = indexfound;
+			if (old_errorcode == err->error_code && err->rdef != NULL) {
```
RIght, this belongs to an else if block. For err, we allocate that before getting into a loop for checking select_index_to_preempt. So it shouldn't be NULL

arungrover commented on src/scheduler/job_info.c:
```diff
@@ -3347,40 +3345,31 @@ find_jobs_to_preempt(status *policy, resource_resv *hjob, server_info *sinfo, in
+				if (old_rdef != err->rdef)
+					filter_again = 1;
+				else {
+					skipto = indexfound;
```
Sure

arungrover commented:
@agrawalravi90 I've addressed your review comments and also moved the code to check for server/queue resources to cull_preemptible_job() function and added a test for it in the PTL script.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -3198,8 +3198,8 @@ find_jobs_to_preempt(status *policy, resource_resv *hjob, server_info *sinfo, in
 	npolicy = nsinfo->policy;
-	njob = find_resource_resv_by_indrank(nsinfo->jobs, hjob->rank, hjob->resresv_ind);
-	prev_prio = njob->job->preempt;
+	hjob_dup = find_resource_resv_by_indrank(nsinfo->jobs, hjob->rank, hjob->resresv_ind);
```
The entire sinfo has been dup'd.  I'd rather not call this hjob_dup.  Maybe hjob?  The first thing I thought of was this was a dup of job in the dup'd server.

arungrover replied:
There is already a variable named hjob in use. It is technically a duplicated high priority job from the duplicate universe. If you don't like it, I can use hjob_new as Ravi suggested. I thought there wasn't anything new with the job that is why choose hjob_dup instead.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -3285,99 +3284,102 @@ find_jobs_to_preempt(status *policy, resource_resv *hjob, server_info *sinfo, in
+			break;
+		}
+
+		else if (old_errorcode == err->error_code && err->rdef != NULL) {
```
Put this else if on the same line as the }

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -3285,99 +3284,102 @@ find_jobs_to_preempt(status *policy, resource_resv *hjob, server_info *sinfo, in
+
+		else if (old_errorcode == err->error_code && err->rdef != NULL) {
+			/* If the error code matches, make sure the resource definition is also matching.
+			 * If the definition does not matche that means the error is because of some other
```
matche=match

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -5161,6 +5146,19 @@ static int cull_running_jobs(resource_resv *job, void *arg)
+			 * RASSN resources are accumulated across the select.
+			 * This means all jobs will have them, and it invalidates this check.
+			 */
+			if (resdef_exists_in_array(policy->resdef_to_check, inp->err->rdef) &&
```
Shouldn't this be resdef_to_check_rassn?  We want to look at flag=q resources that are not flag=qh.

arungrover replied:
Well, actually I think it wouldn't matter because the error is specific to server/queue resource. And we are checking that the resource is not one of the rassn resources that get translated to select-specification. The check was there to make sure that the resource also exists in resdef_to_check list.
Using resdef_to_check_rassn might speed things a little because in most cases it will be a subset of resdef_to_check itself. So it will be a smaller array to search into.
OR, I can remove this check for resdef_to_check completely because if is_ok_to_run is giving a resource definition in error structure then it better be in resdef_to_check.
What do you think?

bhroam replied:
I think you are right, remove the check for resdef_to_check.  You are right that if is_ok_to_run() is returning a resource, it will be in resdef_to_check.

agrawalravi90 commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,480 @@
+        soft/hard limits and also when job can not run because of insufficient
+        resources
+        """
+        a = {'resources_available.ncpus': 4800,
```
Actually I can't really tell the difference between the 2 tests, so putting common code in a separate method will help readers understand what's different

agrawalravi90 commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,480 @@
+        self.server.expect(JOB, {'job_state': 'R'}, id=j1id,
+                           offset=15, interval=5)
+
+        (_, str1) = self.scheduler.log_match(j1id + ";Considering job to run",
```
do you also wanna check for the insufficient resource log message?

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -399,6 +399,81 @@ def test_insufficient_resc_multiple_non_cons(self):
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid_highp, interval=10)
+        self.server.expect(JOB, {ATTR_state: 'S'}, id=jid)
+        search_str = jid_highp + ";Considering job to run"
+        (_, str1) = self.scheduler.log_match(search_str,
```
You are right, but I am just paranoid about timing. If the match is possible then it will find it in the first attempt as well. If you want I can remove max_attempts itself.

agrawalravi90 replied:
The job literally cannot go in the 'R' state before the scheduler logs those messages, so I think it should be fairly safe to make max_attempts = 1.

arungrover replied:
I know, I'm just being irrational here. I'll change it.

arungrover replied:
Well, there isn't a race here as such, since the job is running a log message must also exist. Ravi is right, it isn't needed here.

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,480 @@
+
+    @timeout(3600)
+    @tags('sched', 'scheduling_policy')
+    def test_t1(self):
```
will do

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,480 @@
+        j1.set_sleep_time(9999)
+        j1id = self.server.submit(j1)
+
+        self.server.expect(JOB, {'job_state': 'R'}, id=j1id,
```
You are right, it will be useful.

arungrover commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -0,0 +1,480 @@
+
+    @timeout(3600)
+    @tags('sched', 'scheduling_policy')
+    def test_t2(self):
```
Will do

arungrover commented:
Oh I missed that @agrawalravi90. I've addressed that in the most recent commit

bhroam commented:
Looks good.  I sign off.  Squash and rebase for the merge.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -3360,7 +3360,7 @@ find_jobs_to_preempt(status *policy, resource_resv *hjob, server_info *sinfo, in
 		if (filter_again == 1) {
 			free(rjobs_subset);
-			rjobs_subset = filter_preemptable_jobs(rjobs, rjobs_count, hjob_dup, err);
+			rjobs_subset = filter_preemptable_jobs(rjobs, count_array((void **)rjobs), hjob_dup, err);
```
Remove the parameter and do the count_array() in filter_preemptable_jobs().

arungrover commented:
@bhroam as per our discussion about renaming variable hjob_dup. I have renamed it to nhjob (which is in sync with how we name duplicate universe, nsinfo). Since Ravi is not available to review the code, I got his consent on this change via a mail. Please have a look.

arungrover commented:
Codacy is again not reporting the results. Closing and reopening the PR again

bhroam commented:
I merged in the PR.
